File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
      deque<int>q; vector<int>ans;
       for(int i = 0 ; i<k-1;i++){
          if(q.empty()) q.push_back(i);
          else{ while(!q.empty() && nums[i]>= nums[q.back()]){
            q.pop_back();
           }
           q.push_back(i);}
       }

       for(int i=k-1;i<nums.size();i++){
       
        while(!q.empty() && nums[i]>= nums[q.back()]){
            q.pop_back();
           }
             q.push_back(i);
          
         
        if(q.front()<= (i-k)){ q.pop_front();}
          ans.push_back(nums[q.front()]);
       }
       return ans;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer>dq = new ArrayDeque<>();
         ArrayList<Integer> ls = new ArrayList<>();
         for(int i=0;i<nums.length;i++){
           
           if(!dq.isEmpty() && dq.peekFirst() <= i-k){
            dq.pollFirst();
           }
           while(!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]){
            dq.pollLast();
           }
            dq.addLast(i);
           if(i>=k-1){
             ls.add(nums[dq.peekFirst()]);
           }
         }
         
         int[] ans = new int[ls.size()];
         int idx =0;
        for(int i=0;i<ls.size();i++){
           ans[idx++]= ls.get(i);
        }

        return ans;
    }
}